application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/LicensingSchedulerListener.java

```diff
@@ -0,0 +1,158 @@
+{
+    /**
+     * The role hint of this component.
+     */
+    public static final String ROLE_HINT = "LicensingSchedulerListener";
+
+    /**
+     * The id of application-licensing-licensor-api module..
+     */
+    protected static final String LICENSOR_API_ID = "com.xwiki.licensing:application-licensing-licensor-api";
+
+    protected static final LocalDocumentReference JOB_DOC =
+        new LocalDocumentReference(Arrays.asList("Licenses", "Code"), "AutomaticDependenciesUpgrade");
+
+    private static final List<Event> EVENTS = Arrays.asList(new ExtensionInstalledEvent());
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private WikiDescriptorManager wikiDescriptorManager;
+
+    @Inject
+    private ClassLoaderManager classLoaderManager;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Constructor.
+     */
+    public LicensingSchedulerListener()
+    {
+        super(ROLE_HINT, EVENTS);
+    }
+
+    /**
+     * The unschedule / schedule process should be done at ExtensionUpgradedEvent, but for avoiding
+     * https://jira.xwiki.org/browse/XCOMMONS-751 it is done at initialization step, since when the extension is
+     * initialized after an upgrade, all the extension's listeners are initialized. After the issue is fixed and
+     * licensing starts depending on a version of XWiki >= the version where is fixed, then this code should be moved
+     * inside a ExtensionUpgradedEvent listener.
+     *
+     * @see org.xwiki.component.phase.Initializable#initialize()
+     */
+    @Override
+    public void initialize() throws InitializationException
+    {
+        // Overwrite the Thread Context ClassLoader to work around the https://jira.xwiki.org/browse/XCOMMONS-2064 bug.
+        // Remove this hack once it's fixed and licensing starts depending on XWiki >= the version where it's fixed.
+        Thread.currentThread().setContextClassLoader(
+            new ContextNamespaceURLClassLoader(this.wikiDescriptorManager, this.classLoaderManager));
+
+        try {
+            // Don't trigger the rescheduling process at xwiki startup time.
+            if (this.contextProvider.get() != null) {
+                scheduleAutomaticUpgradesJob(true);
+            }
+        } catch (XWikiException | SchedulerException e) {
+            logger.error("Error while rescheduling AutomaticDependenciesUpgradeJob", e);
+        }
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        String extensionId = ((DefaultInstalledExtension) source).getId().getId();
+
+        if (event instanceof ExtensionInstalledEvent && extensionId.equals(LICENSOR_API_ID)) {
+            try {
+                scheduleAutomaticUpgradesJob(false);
+            } catch (XWikiException | SchedulerException e) {
+                logger.error("Error while scheduling AutomaticDependenciesUpgradeJob after licensing install", e);
```

Comment:
I would throw a exception (RuntimeException) because the observation manager is supposed to catch and handle exceptions from event listeners.